dvc/objects.py

```diff
@@ -0,0 +1,142 @@
+import logging
+
+from .exceptions import DvcException
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectError(DvcException):
+    pass
+
+
+class ObjectFormatError(ObjectError):
+    pass
+
+
+class HashFile:
```

Comment:
A lot of this is inspired by dulwich. E.g. this is a generalized version of ShaFile https://github.com/dulwich/dulwich/blob/35fdccdd4eff2c95137b3d32c0320b805d8f3320/dulwich/objects.py#L256 , since we support many types of hashes and we also support virtual filesystems.